Source/Diffable Collection/FetchedCollectionDiffableDataSource.swift

```diff
@@ -0,0 +1,134 @@
+	}
+
+	// MARK: - Controller management
+
+	/// Sets the delegate of the NSFetchedResultsController instance.
+	private func setDelegate() {
+		controller.delegate = self
+	}
+
+	/// Executes the fetch request of the NSFetchedResultsController instance.
+	private func performFetch() {
+		do {
+			try controller.performFetch()
+		} catch let error {
+			assertionFailure("Error performing controller fetch: \(error)")
+		}
+	}
+
+	// MARK: - NSFetchedResultsControllerDelegate
+	public func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>, didChangeContentWith snapshot: NSDiffableDataSourceSnapshotReference) {
+		var itemsBeforeChange = internalSnapshot.itemIdentifiers
+
+		internalSnapshot.deleteSections(internalSnapshot.sectionIdentifiers)
+		let snapshotWithPermanentIDs = obtainPermanentIDs(for: snapshot, in: controller.managedObjectContext)
+		let typedSnapshot = snapshotWithPermanentIDs as NSDiffableDataSourceSnapshot<String, NSManagedObject>
+		typedSnapshot.sectionIdentifiers.forEach { sectionIdentifier in
+			let section: FetchedDiffableSection = .init(identifier: sectionIdentifier)
+			let items: [FetchedDiffableItem] = typedSnapshot.itemIdentifiers(inSection: sectionIdentifier).map { .init(item: $0) }
+			if (isHidingEmptySections && !items.isEmpty) || !isHidingEmptySections {
+				internalSnapshot.appendSections([section])
+				internalSnapshot.appendItems(items, toSection: section)
+			}
+		}
+
+		var externalSnapshot = internalSnapshot // modifiable snapshot to be displayed
+		delegate?.contentChangeWillBeApplied(snapshot: &externalSnapshot)
+
+		var itemsAfterChange = externalSnapshot.itemIdentifiers
+		var itemsToReconfigure = Array(Set(itemsBeforeChange).intersection(Set(itemsAfterChange)))
+		if #available(iOS 15.0, *), isReconfiguringItems {
+			externalSnapshot.reconfigureItems(itemsToReconfigure)
+		} else {
+			externalSnapshot.reloadItems(itemsToReconfigure)
+		}
+
+		if isUpdatingAutomatically {
+			applySnapshot(snapshot: externalSnapshot)
+		} else {
+			pendingSnapshot = externalSnapshot
+		}
+	}
+
+	// MARK: - Helpers
+	/// The snapshot returned by the `NSFetchedResultsController` instance contains temporary `NSManagedObjectID`s.
+	/// Working with temporary identifiers can lead to issues since at some point in time they will no longer exist.
+	private func obtainPermanentIDs(for snapshot: NSDiffableDataSourceSnapshotReference, in context: NSManagedObjectContext) -> NSDiffableDataSourceSnapshotReference {
+		let typedSnapshot = snapshot as NSDiffableDataSourceSnapshot<String, NSManagedObjectID>
+		let snapshotWithPermanentIDs: NSDiffableDataSourceSnapshotReference = .init()
+		typedSnapshot.sectionIdentifiers.forEach { sectionIdentifier in
+			do {
+				let objects = typedSnapshot.itemIdentifiers(inSection: sectionIdentifier).map { context.object(with: $0) }
+				try context.obtainPermanentIDs(for: objects)
+				snapshotWithPermanentIDs.appendSections(withIdentifiers: [sectionIdentifier])
```

Comment:
Can't we just pre-append all section IDs before the loop? I think it'd be cleaner, and it'd avoid the code duplication (catch block).